File: subprograms/EdgeRDEFromNormalized.py
#!/usr/bin/env python3

"""
EdgeRDEFromNormalized — limma-trend DE on a pre-normalized expression matrix
SalmonStreamer Pipeline Module

Use this when you already have a TMM-normalized expression matrix (or any
matrix on a sample-comparable scale) and you cannot re-run the standard
edgeR quasi-likelihood pipeline because it requires raw integer counts.

Typical use case
----------------
1. Run EdgeRDE on raw counts with --export-normalized-expression to dump
   TMM_normalized_CPM.tsv / TMM_normalized_logCPM.tsv.
2. Post-process the normalized matrix outside the pipeline (e.g. collapse
   paralogs by summing TMM-normalized CPMs).
3. Feed the corrected matrix back into the pipeline via this subcommand to
   produce DE tables and the usual plots.

This wraps src/R_src/edger_de_from_normalized.R, which applies the limma-trend
workflow recommended by the limma user guide for already-normalized data:

    logCPM -> lmFit(design) -> contrasts.fit(contrast) -> eBayes(trend=TRUE)

Metadata format is identical to EdgeRDE: see that module's docstring for
details on --metadata-file vs --group-samples.

@Author: Luis Javier Madrigal-Roca
@Date:   2026-06-02
"""

import argparse
import os
import sys
import subprocess
import tempfile
import csv
import re


def validate_expression_file(filepath):
    """Validate the normalized expression matrix and return sample column names."""
    try:
        with open(filepath, 'r') as f:
            lines = f.readlines()
    except IOError as e:
        raise ValueError(f"Cannot read expression file: {e}")

    if len(lines) < 2:
        raise ValueError(
            "Expression file must have at least 2 rows (header + 1 gene). "
            f"Found {len(lines)} rows."
        )

    header = lines[0].rstrip("\n").split('\t')
    if len(header) < 2:
        raise ValueError(
            "Expression file must have at least 2 columns "
            f"(gene_id + 1 sample). Found {len(header)} columns."
        )

    # Spot-check the first few data rows for numeric values
    for i, line in enumerate(lines[1:5], start=2):
        fields = line.rstrip("\n").split('\t')
        if len(fields) != len(header):
            raise ValueError(
                f"Row {i} has {len(fields)} columns but header has {len(header)}"
            )
        for v in fields[1:]:
            if v == "" or v == "NA":
                continue
            try:
                float(v)
            except ValueError:
                raise ValueError(
                    f"Row {i}, column {fields[0]}: expected numeric value, "
                    f"got '{v}'. (This subcommand expects an already-normalized "
                    "expression matrix, not raw counts.)"
                )

    return header[1:]


def detect_input_scale(filepath):
    """
    Auto-detect whether the matrix is on the log2-CPM scale.

    Heuristic: log-CPM matrices typically contain negative values
    (often around -4 to -8 for very lowly expressed rows after prior_count
    addition) while raw CPM matrices are non-negative. We sample the first
    ~5000 numeric values and return 'logcpm' if any are negative, else 'cpm'.
    """
    sampled = 0
    target = 5000
    with open(filepath, 'r') as f:
        next(f, None)  # header
        for line in f:
            for v in line.rstrip("\n").split('\t')[1:]:
                if v in ("", "NA"):
                    continue
                try:
                    x = float(v)
                except ValueError:
                    continue
                if x < 0:
                    return "logcpm"
                sampled += 1
                if sampled >= target:
                    return "cpm"
    return "cpm"  # default to cpm if file too small / all non-negative


def validate_metadata_file(filepath, sample_names, sample_suffix=None):
    """Same validation as EdgeRDE — check structure and sample overlap."""
    try:
        with open(filepath, 'r') as f:
            lines = f.readlines()
    except IOError as e:
        raise ValueError(f"Cannot read metadata file: {e}")

    if len(lines) < 2:
        raise ValueError("Metadata file must have at least 2 rows (header + 1 sample)")

    header = lines[0].rstrip("\n").split('\t')
    header_lower = [h.lower() for h in header]

    if 'sample_name' not in header_lower:
        raise ValueError(
            "Metadata file must have 'sample_name' column. "
            f"Found columns: {', '.join(header)}"
        )

    has_group = 'group' in header_lower
    has_species_tissue = ('species' in header_lower and 'tissue' in header_lower)
    if not (has_group or has_species_tissue):
        raise ValueError(
            "Metadata file must have either 'group' OR both 'species' and "
            f"'tissue' columns. Found columns: {', '.join(header)}"
        )

    sample_col = header_lower.index('sample_name')
    metadata_samples = set()
    for i, line in enumerate(lines[1:], start=2):
        fields = line.rstrip("\n").split('\t')
        if len(fields) <= sample_col:
            raise ValueError(f"Metadata row {i} has too few columns")
        metadata_samples.add(fields[sample_col])

    stripped = set(sample_names)
    if sample_suffix:
        stripped = set(re.sub(sample_suffix, "", s) for s in sample_names)

    matching = metadata_samples & stripped
    if len(matching) == 0:
        raise ValueError(
            "No samples match between expression file and metadata. "
            f"Expression samples: {', '.join(sorted(list(stripped)[:5]))}... "
            f"Metadata samples: {', '.join(sorted(list(metadata_samples)[:5]))}..."
        )
    pct_matched = 100 * len(matching) / len(stripped)
    if pct_matched < 90:
        print(
            f"WARNING: Only {pct_matched:.1f}% of expression samples match metadata.",
            file=sys.stderr
        )


def build_metadata_from_groups(group_specs):
    rows = []
    for spec in group_specs:
        if ":" not in spec:
            raise ValueError(
                f"--group-samples entry '{spec}' must be in the form "
                "'GroupLabel:sample1,sample2,...'"
            )
        label, samples_str = spec.split(":", 1)
        for s in samples_str.split(","):
            s = s.strip()
            if s:
                rows.append((s, label.strip()))
    return rows


def write_temp_metadata(rows, tmp_dir):
    fh = tempfile.NamedTemporaryFile(
        mode="w", suffix=".tsv", dir=tmp_dir, delete=False
    )
    writer = csv.writer(fh, delimiter="\t")
    writer.writerow(["sample_name", "group"])
    writer.writerows(rows)
    fh.close()
    return fh.name


def main(args):
    r_script = os.path.join(
        os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
        "src", "R_src", "edger_de_from_normalized.R"
    )

    if not os.path.isfile(r_script):
        print(f"ERROR: R script not found at {r_script}", file=sys.stderr)
        sys.exit(1)

    if not os.path.isfile(args.rscript_executable):
        print(
            f"ERROR: Rscript executable not found at: {args.rscript_executable}\n"
            f"       Specify a different path with --rscript-executable",
            file=sys.stderr
        )
        sys.exit(1)

    if args.fdr_threshold < 0 or args.fdr_threshold > 1:
        print(
            f"ERROR: FDR threshold must be between 0 and 1 (got {args.fdr_threshold})",
            file=sys.stderr
        )
        sys.exit(1)
    if args.logfc_threshold < 0:
        print(
            f"ERROR: logFC threshold must be non-negative (got {args.logfc_threshold})",
            file=sys.stderr
        )
        sys.exit(1)
    if args.prior_count <= 0:
        print(
            f"ERROR: --prior-count must be > 0 (got {args.prior_count})",
            file=sys.stderr
        )
        sys.exit(1)

    print("Validating expression file...", file=sys.stderr)
    try:
        sample_names = validate_expression_file(args.expression_file)
    except ValueError as e:
        print(f"ERROR: {e}", file=sys.stderr)
        sys.exit(1)

    # Resolve input scale
    if args.input_scale == "auto":
        detected = detect_input_scale(args.expression_file)
        print(f"Detected input scale: {detected} (override with --input-scale)",
              file=sys.stderr)
        input_scale = detected
    else:
        input_scale = args.input_scale

    # Resolve metadata
    temp_meta_path = None
    if args.metadata_file:
        if not os.path.isfile(args.metadata_file):
            print(f"ERROR: Metadata file not found: {args.metadata_file}",
                  file=sys.stderr)
            sys.exit(1)
        metadata_path = args.metadata_file
    elif args.group_samples:
        try:
            rows = build_metadata_from_groups(args.group_samples)
        except ValueError as e:
            print(f"ERROR: {e}", file=sys.stderr)
            sys.exit(1)
        os.makedirs(args.output_dir, exist_ok=True)
        temp_meta_path = write_temp_metadata(rows, args.output_dir)
        metadata_path = temp_meta_path
        print(f"Built temporary metadata file from --group-samples: {temp_meta_path}")
    else:
        print(
            "ERROR: Provide either --metadata-file or --group-samples.",
            file=sys.stderr
        )
        sys.exit(1)

    if args.metadata_file:
        print("Validating metadata file...", file=sys.stderr)
        try:
            validate_metadata_file(
                metadata_path, sample_names,
                sample_suffix=args.sample_suffix
            )
        except ValueError as e:
            print(f"ERROR: {e}", file=sys.stderr)
            sys.exit(1)

    os.makedirs(args.output_dir, exist_ok=True)

    sample_suffix = args.sample_suffix if args.sample_suffix else "NULL"

    cmd = [
        args.rscript_executable,
        r_script,
        args.expression_file,
        args.output_dir,
        metadata_path,
        str(args.fdr_threshold),
        str(args.logfc_threshold),
        sample_suffix,
        input_scale,
        str(args.prior_count),
    ]

    print("=" * 70)
    print("SalmonStreamer EdgeRDEFromNormalized")
    print("=" * 70)
    print(f"Expression file : {args.expression_file}")
    print(f"Metadata file   : {metadata_path}")
    print(f"Output dir      : {args.output_dir}")
    print(f"Input scale     : {input_scale}")
    if input_scale == "cpm":
        print(f"Prior count     : {args.prior_count}")
    print(f"FDR threshold   : {args.fdr_threshold}")
    print(f"logFC threshold : {args.logfc_threshold}")
    if args.sample_suffix:
        print(f"Sample suffix   : {args.sample_suffix}")
    print()

    try:
        subprocess.run(cmd, check=True, capture_output=False)
    except subprocess.CalledProcessError as e:
        print(
            f"ERROR: R script failed with exit code {e.returncode}",
            file=sys.stderr
        )
        sys.exit(e.returncode)
    except FileNotFoundError as e:
        print(
            f"ERROR: Cannot find Rscript executable: {args.rscript_executable}\n"
            f"       {e}",
            file=sys.stderr
        )
        sys.exit(1)
    finally:
        if temp_meta_path and os.path.isfile(temp_meta_path):
            os.remove(temp_meta_path)

    print("\nEdgeRDEFromNormalized complete.")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description=(
            "Run limma-trend differential expression on a pre-normalized "
            "expression matrix (SalmonStreamer EdgeRDEFromNormalized subcommand)."
        ),
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Typical workflow:
  1. SalmonStreamer.py EdgeRDE ... --export-normalized-expression
  2. (collapse paralogs / curate the TMM_normalized matrix outside the pipeline)
  3. SalmonStreamer.py EdgeRDEFromNormalized \\
       --expression-file TMM_normalized_logCPM.collapsed.tsv \\
       --metadata-file metadata.tsv \\
       --output-dir DE_results_collapsed/

Input scale:
  --input-scale logcpm  -> values used as-is (default if file contains
                            negative values, e.g. our TMM_normalized_logCPM.tsv)
  --input-scale cpm     -> log2(x + --prior-count) applied internally
                            (e.g. our TMM_normalized_CPM.tsv with prior 0.25)
  --input-scale auto    -> detect from the matrix (negative values -> logcpm)
        """
    )

    parser.add_argument(
        "--expression-file", required=True,
        help="Tab-separated normalized expression matrix (genes x samples, "
             "first column = TranscriptID)."
    )
    parser.add_argument(
        "--output-dir", required=True,
        help="Directory for all output files."
    )
    parser.add_argument(
        "--metadata-file", default=None,
        help=(
            "Tab-separated sample metadata. Required columns: sample_name, group "
            "(or species + tissue). Takes priority over --group-samples."
        )
    )
    parser.add_argument(
        "--group-samples", nargs="+", default=None, metavar="GROUP:s1,s2,...",
        help=(
            "Inline group specification when no metadata file is available. "
            "Format: 'GroupLabel:sample1,sample2,...' (one entry per group)."
        )
    )
    parser.add_argument(
        "--input-scale", choices=["auto", "logcpm", "cpm"], default="auto",
        help=(
            "Scale of the input matrix. 'logcpm' = use as-is; "
            "'cpm' = apply log2(x + prior_count); "
            "'auto' = detect from the data (default)."
        )
    )
    parser.add_argument(
        "--prior-count", type=float, default=0.25,
        help="Pseudo-count added before log2 when --input-scale is 'cpm' (default: 0.25)."
    )
    parser.add_argument(
        "--fdr-threshold", type=float, default=0.05,
        help="FDR significance cutoff (default: 0.05)."
    )
    parser.add_argument(
        "--logfc-threshold", type=float, default=1.0,
        help="|log2FC| threshold for significance categories (default: 1.0)."
    )
    parser.add_argument(
        "--sample-suffix", default=None,
        help=(
            "Regex pattern stripped from expression matrix column names before "
            "matching to metadata sample_name values (e.g. '_R1_filtered$'). Optional."
        )
    )
    parser.add_argument(
        "--rscript-executable",
        default=os.path.expanduser("~/.conda/envs/PyR/bin/Rscript"),
        help="Path to the Rscript executable (default: ~/.conda/envs/PyR/bin/Rscript)."
    )

    main(parser.parse_args())

File: src/R_src/edger_de_from_normalized.R
#!/usr/bin/env Rscript

################################################################################
# Differential Expression from a pre-normalized expression matrix
# SalmonStreamer Pipeline Module (limma-trend pipeline)
#
# Author: Luis Javier Madrigal-Roca
# Date:   2026-06-02
#
# Use this when the input matrix has already been TMM-normalized (or otherwise
# normalized to a sample-comparable scale) and you need to run DE on the
# normalized values — for example, after collapsing paralogs by summing
# TMM-normalized CPMs.
#
# The classic edgeR quasi-likelihood pipeline cannot be reused here because it
# requires raw integer counts. Instead we apply the limma-trend workflow:
#
#   logCPM -> lmFit(design) -> contrasts.fit(contrast) -> eBayes(trend=TRUE)
#
# This is the workflow limma's user guide recommends for already-normalized
# expression data (Section "When to use limma-trend").
#
# Usage (called by SalmonStreamer EdgeRDEFromNormalized):
#   Rscript edger_de_from_normalized.R <input_file> <output_dir> <metadata_file> \
#                                      [fdr_threshold] [logfc_threshold] \
#                                      [sample_suffix] [input_scale] [prior_count]
#
# Arguments:
#   input_file       Tab-separated normalized expression matrix
#                    (genes/transcripts x samples). First column = TranscriptID.
#
#   output_dir       Directory where all output files are written.
#
#   metadata_file    Same format as edger_de.R: requires `sample_name` and
#                    `group` columns (or `species` + `tissue`).
#
#   fdr_threshold    FDR significance cutoff (default: 0.05).
#   logfc_threshold  |log2FC| threshold for significance categories (default: 1).
#   sample_suffix    Regex stripped from column names before matching to
#                    metadata sample_name values. Pass "NULL" to skip.
#   input_scale      "logcpm" (default) or "cpm".
#                    - logcpm: values are used as-is for lmFit.
#                    - cpm   : values are converted via log2(x + prior_count)
#                              before fitting.
#   prior_count      Pseudo-count added before log2 when input_scale == "cpm"
#                    (default: 0.25).
################################################################################

options(bitmapType = "cairo")
if (!interactive()) options(device = "pdf")

suppressPackageStartupMessages({
  library(limma)
  library(ggplot2)
  library(pheatmap)
  library(RColorBrewer)
  library(dplyr)
  library(readr)
})

# ---------------------------------------------------------------------------
# 0. Argument parsing
# ---------------------------------------------------------------------------

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 3) {
  stop(
    "Usage: edger_de_from_normalized.R <input_file> <output_dir> <metadata_file> ",
    "[fdr_threshold] [logfc_threshold] [sample_suffix] [input_scale] [prior_count]"
  )
}

input_file      <- args[1]
output_dir      <- args[2]
metadata_arg    <- args[3]
fdr_threshold   <- if (length(args) >= 4) as.numeric(args[4]) else 0.05
logfc_threshold <- if (length(args) >= 5) as.numeric(args[5]) else 1.0
sample_suffix   <- if (length(args) >= 6 && args[6] != "NULL") args[6] else NULL
input_scale     <- if (length(args) >= 7) tolower(args[7]) else "logcpm"
prior_count     <- if (length(args) >= 8) as.numeric(args[8]) else 0.25

if (!input_scale %in% c("logcpm", "cpm")) {
  stop("input_scale must be 'logcpm' or 'cpm' (got '", input_scale, "')")
}

metadata_file <- if (metadata_arg == "NULL") NULL else metadata_arg

dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

cat("=== SalmonStreamer EdgeRDEFromNormalized Analysis ===\n")
cat("Input file     :", input_file, "\n")
cat("Output dir     :", output_dir, "\n")
cat("Metadata file  :", ifelse(is.null(metadata_file), "(none)", metadata_file), "\n")
cat("Input scale    :", input_scale, "\n")
if (input_scale == "cpm") {
  cat("Prior count    :", prior_count, "\n")
}
cat("FDR threshold  :", fdr_threshold, "\n")
cat("logFC threshold:", logfc_threshold, "\n\n")

# ---------------------------------------------------------------------------
# 1. Load expression matrix
# ---------------------------------------------------------------------------

if (!file.exists(input_file)) stop("Input file not found: ", input_file)

expr_data      <- read_tsv(input_file, show_col_types = FALSE)
transcript_ids <- expr_data[[1]]
expr_matrix    <- as.matrix(expr_data[, -1])
rownames(expr_matrix) <- transcript_ids
storage.mode(expr_matrix) <- "double"

cat("Data dimensions:", nrow(expr_matrix), "genes x",
    ncol(expr_matrix), "samples\n")

if (!is.null(sample_suffix)) {
  colnames(expr_matrix) <- gsub(sample_suffix, "", colnames(expr_matrix))
  cat("Stripped suffix '", sample_suffix, "' from sample names\n", sep = "")
}
sample_names <- colnames(expr_matrix)

# Convert raw CPM to log2-CPM with prior
if (input_scale == "cpm") {
  if (any(expr_matrix < 0, na.rm = TRUE)) {
    stop(
      "Input was declared as raw CPM (input_scale='cpm') but contains ",
      "negative values. Did you mean input_scale='logcpm'?"
    )
  }
  cat("Converting raw CPM -> log2(CPM + ", prior_count, ")\n", sep = "")
  expr_matrix <- log2(expr_matrix + prior_count)
}

logcpm <- expr_matrix  # from here on, everything assumes log-scale

# ---------------------------------------------------------------------------
# 2. Load and validate metadata
# ---------------------------------------------------------------------------

if (is.null(metadata_file)) {
  stop(
    "A metadata file is required. Provide it via --metadata-file or ",
    "use --group-samples to let the Python wrapper build one automatically."
  )
}
if (!file.exists(metadata_file)) stop("Metadata file not found: ", metadata_file)

metadata <- read_tsv(metadata_file, show_col_types = FALSE)

col_lower <- tolower(colnames(metadata))
if ("sample_name" %in% col_lower) {
  colnames(metadata)[col_lower == "sample_name"] <- "Sample"
} else if (!"sample" %in% col_lower) {
  stop("Metadata must contain a 'sample_name' or 'Sample' column.")
} else {
  colnames(metadata)[col_lower == "sample"] <- "Sample"
}

col_lower <- tolower(colnames(metadata))
if ("group" %in% col_lower) {
  colnames(metadata)[col_lower == "group"] <- "Group"
} else if (all(c("species", "tissue") %in% col_lower)) {
  colnames(metadata)[col_lower == "species"] <- "Species"
  colnames(metadata)[col_lower == "tissue"]  <- "Tissue"
  metadata$Group <- paste(metadata$Species, metadata$Tissue, sep = "_")
  cat("Derived Group from species + tissue columns\n")
} else {
  stop(
    "Metadata must contain either a 'group' column or both ",
    "'species' and 'tissue' columns."
  )
}

in_both   <- intersect(metadata$Sample, sample_names)
only_meta <- setdiff(metadata$Sample, sample_names)
only_mat  <- setdiff(sample_names, metadata$Sample)

if (length(only_meta) > 0)
  warning("Samples in metadata but not in expression matrix: ",
          paste(only_meta, collapse = ", "))
if (length(only_mat) > 0)
  warning("Samples in expression matrix but not in metadata: ",
          paste(only_mat, collapse = ", "))
if (length(in_both) == 0)
  stop("No samples shared between metadata and expression matrix. ",
       "Check sample_name values and any --sample-suffix stripping.")

metadata <- metadata[metadata$Sample %in% in_both, ]
logcpm   <- logcpm[, metadata$Sample, drop = FALSE]

cat("Samples retained:", nrow(metadata), "\n")
cat("Groups          :", paste(unique(metadata$Group), collapse = ", "), "\n\n")

# ---------------------------------------------------------------------------
# 3. Design matrix
# ---------------------------------------------------------------------------

metadata$group_factor <- factor(metadata$Group)
design <- model.matrix(~ 0 + group_factor, data = metadata)
colnames(design) <- make.names(levels(metadata$group_factor))

# ---------------------------------------------------------------------------
# 4. Exploratory plots (PCA + sample correlation)
# ---------------------------------------------------------------------------

cat("Generating exploratory plots...\n")

# Drop rows that are constant or all-NA before PCA / correlation
row_sd <- apply(logcpm, 1, sd, na.rm = TRUE)
keep_var <- is.finite(row_sd) & row_sd > 0
if (sum(keep_var) < nrow(logcpm)) {
  cat("Dropping", nrow(logcpm) - sum(keep_var),
      "constant / all-NA rows from PCA & correlation plots\n")
}
logcpm_var <- logcpm[keep_var, , drop = FALSE]

pca_result <- prcomp(t(logcpm_var), scale. = FALSE)
pct_var    <- round(summary(pca_result)$importance[2, 1:2] * 100, 1)

pca_df <- data.frame(
  PC1    = pca_result$x[, 1],
  PC2    = pca_result$x[, 2],
  Sample = rownames(pca_result$x)
)
pca_df <- merge(pca_df, metadata[, c("Sample", "Group")], by = "Sample")
hulls  <- pca_df %>% group_by(Group) %>% slice(chull(PC1, PC2))

p_pca <- ggplot(pca_df, aes(x = PC1, y = PC2, color = Group, fill = Group)) +
  geom_polygon(data = hulls, alpha = 0.15, show.legend = FALSE) +
  geom_point(size = 3, alpha = 0.85) +
  labs(
    title = "PCA of Normalized Expression",
    x     = paste0("PC1 (", pct_var[1], "%)"),
    y     = paste0("PC2 (", pct_var[2], "%)")
  ) +
  theme_bw() +
  theme(legend.title = element_blank(), panel.grid = element_blank())

ggsave(file.path(output_dir, "PCA_plot.pdf"), p_pca,
       width = 8, height = 6, device = "pdf")
ggsave(file.path(output_dir, "PCA_plot.png"), p_pca,
       width = 8, height = 6, dpi = 300, device = "png")

cor_matrix <- cor(logcpm_var, method = "pearson")
ann_df     <- data.frame(Group = metadata$Group, row.names = metadata$Sample)

pdf(file.path(output_dir, "sample_correlation_heatmap.pdf"),
    width = 10, height = 10)
pheatmap(cor_matrix, annotation_col = ann_df, annotation_row = ann_df,
         main = "Sample Correlation Heatmap")
dev.off()

# ---------------------------------------------------------------------------
# 5. limma-trend fit
# ---------------------------------------------------------------------------

cat("Fitting limma-trend model on log-CPM...\n")
fit_full <- lmFit(logcpm, design)

# ---------------------------------------------------------------------------
# 6. All pairwise contrasts + per-contrast topTable
# ---------------------------------------------------------------------------

groups      <- colnames(design)
orig_groups <- levels(metadata$group_factor)
cat("Groups:", paste(orig_groups, collapse = ", "), "\n\n")

results_list  <- list()
all_sig_genes <- character(0)

for (i in seq_len(length(groups) - 1)) {
  for (j in (i + 1):length(groups)) {
    g1     <- groups[i]
    g2     <- groups[j]
    g1_lab <- orig_groups[i]
    g2_lab <- orig_groups[j]
    cmp    <- paste0(g1_lab, "_vs_", g2_lab)

    contrast_vec     <- setNames(numeric(length(groups)), groups)
    contrast_vec[g1] <- 1
    contrast_vec[g2] <- -1

    fit_c <- contrasts.fit(fit_full, contrasts = matrix(contrast_vec, ncol = 1))
    fit_c <- eBayes(fit_c, trend = TRUE)
    tt    <- topTable(fit_c, number = Inf, sort.by = "P")

    # Normalise output column names to match the EdgeRDE schema where possible
    if ("adj.P.Val" %in% colnames(tt)) {
      colnames(tt)[colnames(tt) == "adj.P.Val"] <- "FDR"
    }
    if ("P.Value" %in% colnames(tt)) {
      colnames(tt)[colnames(tt) == "P.Value"] <- "PValue"
    }
    if ("AveExpr" %in% colnames(tt)) {
      colnames(tt)[colnames(tt) == "AveExpr"] <- "logCPM"
    }

    tt$TranscriptID <- rownames(tt)
    tt <- tt[, c("TranscriptID", setdiff(colnames(tt), "TranscriptID"))]
    tt$Significance <- "Not Significant"
    tt$Significance[tt$FDR < fdr_threshold &
                      abs(tt$logFC) > logfc_threshold] <- "Significant"
    tt$Significance[tt$FDR < 0.01 &
                      abs(tt$logFC) > 2 * logfc_threshold] <- "Highly Significant"
    tt <- tt[order(tt$FDR), ]

    results_list[[cmp]] <- tt

    write_tsv(tt, file.path(output_dir, paste0(cmp, "_DE_results.tsv")))

    sig <- tt[tt$FDR < fdr_threshold, ]
    write_tsv(sig, file.path(output_dir, paste0(cmp, "_significant_genes.tsv")))
    all_sig_genes <- union(all_sig_genes, sig$TranscriptID)

    cat(sprintf("  %s -> tested: %d | sig: %d | up: %d | down: %d\n",
                cmp, nrow(tt), nrow(sig),
                sum(sig$logFC > 0), sum(sig$logFC < 0)))

    # Volcano plot
    vdata <- tt
    vdata$neg_log10_FDR <- -log10(pmax(vdata$FDR, .Machine$double.xmin))

    p_vol <- ggplot(vdata, aes(x = logFC, y = neg_log10_FDR,
                               color = Significance)) +
      geom_point(alpha = 0.5, size = 0.8) +
      scale_color_manual(values = c(
        "Not Significant"    = "grey70",
        "Significant"        = "#377EB8",
        "Highly Significant" = "#E41A1C"
      )) +
      geom_vline(xintercept = c(-logfc_threshold, logfc_threshold),
                 linetype = "dashed", alpha = 0.5) +
      geom_hline(yintercept = -log10(fdr_threshold),
                 linetype = "dashed", alpha = 0.5) +
      labs(
        title = paste("Volcano:", cmp),
        x     = "log2 Fold Change",
        y     = "-log10(FDR)"
      ) +
      theme_bw() +
      theme(legend.title = element_blank(), panel.grid = element_blank())

    ggsave(file.path(output_dir, paste0(cmp, "_volcano.pdf")), p_vol,
           width = 10, height = 8, device = "pdf")
    ggsave(file.path(output_dir, paste0(cmp, "_volcano.png")), p_vol,
           width = 10, height = 8, dpi = 300, device = "png")
  }
}

# Heatmap of top DE genes
if (length(all_sig_genes) > 0) {
  genes_to_plot <- if (length(all_sig_genes) > 100) {
    unique(unlist(lapply(results_list, function(x)
      head(x$TranscriptID[x$FDR < fdr_threshold], 50))))
  } else {
    all_sig_genes
  }
  genes_to_plot <- intersect(genes_to_plot, rownames(logcpm))

  if (length(genes_to_plot) > 0) {
    hz <- t(scale(t(logcpm[genes_to_plot, , drop = FALSE])))
    pdf(file.path(output_dir, "DE_genes_heatmap.pdf"),
        width = 12, height = max(8, length(genes_to_plot) * 0.15))
    pheatmap(hz, annotation_col = ann_df,
             show_rownames = length(genes_to_plot) <= 50,
             main = "Top Differentially Expressed Genes",
             fontsize_row = 6)
    dev.off()
  }
}

# ---------------------------------------------------------------------------
# 7. Summary report
# ---------------------------------------------------------------------------

summary_lines <- c(
  "=== SalmonStreamer EdgeRDEFromNormalized Analysis Summary ===",
  paste("Date            :", Sys.Date()),
  paste("Input file      :", input_file),
  paste("Metadata        :", metadata_file),
  paste("Input scale     :", input_scale),
  if (input_scale == "cpm") paste("Prior count     :", prior_count) else NULL,
  paste("Genes (total)   :", nrow(logcpm)),
  paste("Samples         :", ncol(logcpm)),
  paste("Groups          :", paste(orig_groups, collapse = ", ")),
  paste("FDR threshold   :", fdr_threshold),
  paste("logFC threshold :", logfc_threshold),
  "",
  "Comparisons:"
)
for (cmp in names(results_list)) {
  tbl <- results_list[[cmp]]
  sig <- sum(tbl$FDR < fdr_threshold, na.rm = TRUE)
  summary_lines <- c(summary_lines,
    sprintf("  %-60s %d significant genes", cmp, sig))
}

writeLines(summary_lines, file.path(output_dir, "analysis_summary.txt"))
writeLines(capture.output(sessionInfo()),
           file.path(output_dir, "session_info.txt"))

cat("\nDone. Results written to:", output_dir, "\n")
